Add CLI

jacoblearned opened the pull request:
```
pcf -h
Usage: pcf [--version] [--help] <command> [<args>...]

  Particle Cloud Framework

Options:
  -v, --version  Show PCF version.
  -h, --help     Show this message and exit.

Commands:
  apply      Set a desired state and apply changes
  run        Set desired state to 'run' and apply changes
  stop       Set desired state to 'stopped' and apply changes
  terminate  Set desired state to 'terminated' and apply changes
```

#### Changes
Adds a [Click](http://click.palletsprojects.com/en/7.x/)-based CLI, `pcf`, with four commands (`apply`, `run`, `stop`, and `terminate`) for applying state changes to Particles and Quasiparticles
* CLI has autosuggestion for close typos
* Respects a user's `NO_COLOR` env var if set. See https://no-color.org

Adds a `test` Make task that will execute `pytest` and generate code coverage reports with uncovered line numbers.
* Adds a `.coveragerc` file to ignore the `pcf/test` directory in coverage reports

##### Overview
The CLI will by default look for the following config files in order of their precedence:
1. `pcf.json`
2. `pcf.yml`
3. `pcf.yaml`

Users can provide their own custom-named files with the `-f` or `--file` option.

These PCF config files can contain a single dict for a single particle definition or a list of particle definition dictionaries.

Example `pcf.yml` file:
```yaml
---
pcf_name: my_ec2
flavor: ec2_instance
aws_resource:
  custom_config:
    instance_name: my-ec2
    ...
```

The Particle/Quasiparticle lookup method searches for the corresponding class object that has the flavor given by the user set as a class variable since they are unique.


##### Commands
`pcf apply` is a general command that takes a `--state` or `-s` option (`running` as default), sets desired state, and runs `apply()` on the particle/quasiparticle class
* Example: `pcf apply -s terminated my_ec2`

`pcf run <pcf_name>` is a convenience command equivalent to `pcf apply --state running <pcf_name>`

`pcf stop <pcf_name>` is a convenience command equivalent to `pcf apply --state stopped <pcf_name>`

`pcf terminate <pcf_name>` is a convenience command equivalent to `pcf apply --state terminate <pcf_name>`

All four applying commands take a `-c` or `--cascade` flag to determine if state changes should cascade through family members as well as a `-q` or `--quiet` flag to prevent CLI output except when errors occur.

##### Tests
* Adds `conftest.py` with reusable pytest fixtures for a [CLIRunner](http://click.palletsprojects.com/en/7.x/api/#click.testing.CliRunner) and a convenient method to copy test config files into a temp directory.
* Adds tests for commands and all utility functions

To test the CLI, check out this branch, run `pip install -e .` and run `pcf --version` to get started.

jacoblearned committed:
gitignore venv directories from virtualenv

jacoblearned committed:
Add color and failure utils for CLI

jacoblearned committed:
Update to boto3 v1.9.76

This prevents a version conflict with the version of botocore we were using

jacoblearned committed:
Move pcf version to importable variable

jacoblearned committed:
Add top-level pcf command

jacoblearned committed:
Add suggestion for command typos

Uses Levenshtein distance against list of supported commands specified as PCFCLI class variable

jacoblearned committed:
Add list_commands method to PCFCLI

This is used by Click to list the commands in help prompts

jacoblearned committed:
Add apply command placeholder

jacoblearned committed:
Add help when just 'pcf' is the command

jacoblearned committed:
Add tests for top-level pcf command

jacoblearned committed:
Update awscli to 1.16.84

jacoblearned committed:
Remove unused module in CLI

jacoblearned committed:
Add pyyaml to required packages

jacoblearned committed:
Add did_you_mean utility function

Helps with printing potentially intended string values

jacoblearned committed:
Add class-finding utilty from a particle flavor

jacoblearned committed:
Add utility functions for reading pcf config files

jacoblearned committed:
Moves compute client to constructor of ComputeEngine

The previous position will execute the Google Authentication Code, which may cause unexpected errors when performing metaprogramming for tasks such as loading the class from the CLI

jacoblearned committed:
Fix particle class search and clean error msgs

jacoblearned committed:
Normalize pcf config to list when reading from file

jacoblearned committed:
Add Particle class instantiation to CLI

Takes a target pcf_name value and searches a config file for the specified particle config, suggesting potentially close typos if the particle class or pcf_name isn't found

jacoblearned committed:
Add basic apply command execution

Attempts to load the particle class, set the desired state based on the value of the state flag (defaulting to State.running) and then runs apply on the particle

jacoblearned committed:
Add pcf run command

Convenience comamnd equivalent to
`pcf apply my_ec2 --state running`

jacoblearned committed:
Add pcf stop command

jacoblearned committed:
Add pcf terminate command

jacoblearned committed:
Add utility function for executing an applying command

jacoblearned committed:
Add pytest fixture for copying test PCF config files

jacoblearned committed:
Add test pcf config files

jacoblearned committed:
Add tests for apply command

jacoblearned committed:
Add tests for run command

jacoblearned committed:
Add tests for stop command

jacoblearned committed:
Add tests for terminate command

jacoblearned committed:
Format command test code

jacoblearned committed:
Format run, stop, and terminate commands

jacoblearned committed:
Improve help messages for commands

jacoblearned committed:
Format stop and terminate commands

jacoblearned committed:
Combine no_color and color utility functions

jacoblearned committed:
Compare levenshtein distance with lowercase strings

jacoblearned committed:
Add tests for color, fail, and similar_strings util functions

jacoblearned committed:
Use basename for output in when loading pcf config

jacoblearned committed:
Simplify pcf config file loading interface

jacoblearned committed:
Add ill-formed pcf config file fixtures

jacoblearned committed:
Improve error msgs when loading particles from file

jacoblearned committed:
Add useful test deployments to pcf.yml

jacoblearned committed:
Change silent option to quiet

jacoblearned committed:
Add -q or --quiet flag to apply, run, stop, and terminate

jacoblearned committed:
Add tests for utility functions

jacoblearned committed:
Add -c or --cascade option for applying with cascade=True

jacoblearned committed:
Scope pytest testpath to pcf/test directory

jacoblearned committed:
Handle import errors when searching for particle class

jacoblearned committed:
Add test for import errors in submodule lookup

jacoblearned committed:
Add make test task w/ code coverage

Adds a .coveragerc to ignore the tests/directory when generating coverage reports

jacoblearned committed:
Add tests for missing LoC

Tests bare `pcf` command and non-dict entries in a pcf config file list

anovis commented:
looks really good. just a few things right now.

can we add a state command that returns the status of the particle/quasiparticle

after apply finishes

instead of` Successfully applied changes to ec2-example` maybe return the state

dont think commands should require `pcf_name` field. it could be optional if you only want to run one particle or a list of particles otherwise it will run everything in the json def

anovis commented:
`Applying changes to ec2-example...` might want to display the pending state every interval... 
for apply can we add the max timeout parameter as well.

anovis commented:
`Loading Particle/Quasiparticle flavor for ec2-example...`

should we display the actual flavor that it is loading?

anovis commented on pcf/cli/commands/run.py:
```diff
@@ -0,0 +1,41 @@
+from pcf.cli.utils import execute_applying_command
+
+
+@click.command(name="run", short_help="Set desired state to 'run' and apply changes")
```
should be state to `running`

anovis commented on pcf/cli/commands/terminate.py:
```diff
@@ -0,0 +1,45 @@
+    show_default=True,
+    help="The JSON or YAML file defining your infrastructure configuration",
+)
+@click.option(
```
anyway to share these options so you dont have to rewrite them every-time
maybe using decorators?

https://stackoverflow.com/questions/40182157/python-click-shared-options-and-flags-between-commands

jacoblearned replied:
Yeah that's a good call.

anovis commented on pcf/cli/utils.py:
```diff
@@ -0,0 +1,238 @@
+
+
+def particle_class_from_flavor(flavor):
+    """ Return the class object of the given flavor (or None) by searching
```
would this work if the user has custom defined flavors not in the core... maybe env var with location or something in the config to specify.

jacoblearned replied:
Not this function. I wasn't aware of a formal process for building custom Particles external to the package. Is this documented anywhere?

anovis replied:
https://capitalone.github.io/Particle-Cloud-Framework/docs/build/html/sections/quickstartdeveloper.html

you would use the same way, but would import from a different path

jacoblearned replied:
That doc recommends pull requests into this package to integrate custom-built Particles and Quasiparticles. Are you suggesting that the CLI should support an optional field in the config file per particle deployment in which users have to specify Python module path strings for this function to look for like this:

```yaml
pcf_name: my_custom_particle
flavor: my_custom_particle_flavor
external_package: my_module.my_custom_particle
aws_resources:
   ...
```

And then the CLI strip out the external_package field before passing the config to the Particle constructor if it finds the right class in `my_module.my_custom_particle`?

anovis commented on pcf/cli/utils.py:
```diff
@@ -0,0 +1,238 @@
+    return submodules
+
+
+def particle_class_from_flavor(flavor):
```
this should probably go into the core pcf utils. might add to the particle / quasiparticle class itself later on

anovis commented on pcf/cli/utils.py:
```diff
@@ -0,0 +1,238 @@
+    if not quiet:
+        click.secho("Applying changes to {0}...".format(pcf_name), fg=color("blue"))
+
+    particle.apply(cascade=cascade)
```
add max timeout params here as well.

jacoblearned commented:
@anovis Since this PR is quite large, I was planning on adding the state command in a subsequent PR. I agree with the preference to show the state after apply finishes as well and will include that too.

jacoblearned commented:
> `Applying changes to ec2-example...` might want to display the pending state every interval...
> for apply can we add the max timeout parameter as well.

What interval are you referring to?

jacoblearned committed:
Add click_options util to extract common CLI options

jacoblearned committed:
Add max timeout common CLI option

jacoblearned committed:
Move particle_class_from_flavor to pcf core utils

jacoblearned committed:
Include flavor name in loading message

jacoblearned committed:
Make pcf_name an optional argument

If the pcf_name argument isn't given to an applying command, all apply() will be called on all particles

jacoblearned commented:
@anovis I pushed some changes for the suggestions. Let me know what you think whenever you have the time 👍

anovis commented:
there is no set interval now so it would be up to you for what makes sense. this was just a suggestion, but some particles can take awhile so it might be good for the user to see that it is in a pending state.

anovis commented on pcf/cli/commands/__init__.py:
```diff
@@ -0,0 +1,31 @@
+import click
+
+COMMON_OPTIONS = [
```
these are just common options for the apply commands. so maybe change to apply_options or common_apply_options.

anovis commented on pcf/cli/commands/__init__.py:
```diff
@@ -0,0 +1,31 @@
+        "-t",
+        "--timeout",
+        type=int,
+        help="The maximum number of seconds to wait before a timeout error is thrown",
```
default is None

jacoblearned replied:
If the flag isn't passed by the user it will be `None` by default.

anovis commented on pcf/cli/commands/stop.py:
```diff
@@ -0,0 +1,29 @@
+
+
+@click.command(
+    name="stop", short_help="Set desired state to 'stopped' and apply changes"
```
just apply not `apply changes`

anovis commented on pcf/cli/commands/terminate.py:
```diff
@@ -0,0 +1,29 @@
+
+
+@click.command(
+    name="terminate", short_help="Set desired state to 'terminated' and apply changes"
```
just apply not `apply changes`

anovis commented on pcf/cli/commands/terminate.py:
```diff
@@ -0,0 +1,29 @@
+@click.argument("pcf_name", required=False)
+@click.pass_context
+def terminate(ctx, pcf_name, cascade, quiet, file_, timeout):
+    """ Set desired state to 'terminated' and apply changes
```
just apply not `apply changes`

jacoblearned committed:
Clean applying cmd options and help messages

jacoblearned committed:
Add progress bar for applies to >1 particle

jacoblearned committed:
Remove "changes" from command help strs

jacoblearned commented:
@anovis I added a progress bar for when `apply` runs on multiple particles.

anovis commented:
is there a better way to install the cli for a user than running `pip install -e .`

anovis commented:
can you add a readme page or add a section to the docs? even just the screenshot you have in the pr works!

jacoblearned commented:
> is there a better way to install the cli for a user than running `pip install -e .`

`pip install -e .` is just for testing the CLI in a development environment. It creates an editable install as defined by the root `setup.py`. 

A regular user would just run `pip install pcf` and then have access to the `pcf` CLI because it is defined as a [`console_script` entrypoint in `setup.py`](https://github.com/capitalone/Particle-Cloud-Framework/blob/add-apply-command/setup.py#L35-L38)

davidyum commented:
awesome thanks Jacob! Will review and try this out today

davidyum commented on pcf/__init__.py:
```diff
@@ -0,0 +1,3 @@
+import os
+
+VERSION = os.environ.get('PCF_TAG', '0.1.1')
```
I think we are at pcf 0.3.2 now.

davidyum commented on pcf/cli/commands/apply.py:
```diff
@@ -0,0 +1,34 @@
+    "-s",
+    "--state",
+    type=click.Choice(["running", "stopped", "terminated"], case_sensitive=False),
+    default="running",
```
we might want to force them to specify desired state instead of defaulting to running for consistency.

davidyum reviewed (commented):
looks good! Tested it with aws/gcp particles and quasiparticles and all worked smoothly.

davidyum commented:
Anything like a README or how to get started guide for users would be helpful. Almost just like the first comment you have at the top :)

jacoblearned committed:
Update default version to  0.3.2

jacoblearned committed:
Add doc for CLI usage

davidyum commented:
@jacoblearned theres a merge conflict with pcf/particle/gcp/compute/compute.py because we renamed but I think we are good to merge once we have that fixed.

jacoblearned commented:
@davidyum Resolved the conflict. Thanks for reminding me 👍
